File: sql/src/main/java/de/mel/sql/ISQLResource.java
package de.mel.sql;

import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by xor on 2/6/17.
 */
public interface ISQLResource<T extends SQLTableObject> extends AutoCloseable {
    T getNext() throws SqlQueriesException;

    @Override
    void close() throws SqlQueriesException;

    boolean isClosed() throws SqlQueriesException;

    default public List<T> toList() throws SqlQueriesException {
        List<T> list = new ArrayList<>();
        T item = getNext();
        while (item != null) {
            list.add(item);
            item = getNext();
        }
        return list;
    }
}

File: sql/src/main/java/de/mel/sql/ISQLResource.kt
package de.mel.sql

import de.mel.sql.SqlQueriesException
import java.util.*

/**
 * Created by xor on 2/6/17.
 */
interface ISQLResource<T : SQLTableObject?> : AutoCloseable {
    @get:Throws(SqlQueriesException::class)
    val next: T?

    fun next() = next

    @Throws(SqlQueriesException::class)
    override fun close()

    @get:Throws(SqlQueriesException::class)
    val isClosed: Boolean

    @Throws(SqlQueriesException::class)
    fun toList(): List<T> {
        val list: MutableList<T> = ArrayList()
        var item: T? = next
        while (item != null) {
            list.add(item)
            item = next
        }
        return list
    }
}